library/alloc/src/collections/vec_deque/into_iter.rs

```diff
@@ -52,6 +52,126 @@ impl<T, A: Allocator> Iterator for IntoIter<T, A> {
         let len = self.inner.len();
         (len, Some(len))
     }
+
+    #[inline]
+    fn advance_by(&mut self, n: usize) -> Result<(), usize> {
+        if self.inner.len < n {
+            let len = self.inner.len;
+            self.inner.clear();
+            Err(len)
+        } else {
+            self.inner.drain(..n);
```

Comment:
The drain machinery is quite complicated. A more direct implementation such as updating the offset/len fields and doing 2 `drop_in_place` calls should be possible here.

Reply:
I originally used something along the lines of this:
```rs
let (a, b) = self.inner.slice_ranges(..n);
let (a, b) = (self.inner.buffer_range(a), self.inner.buffer_range(b));
self.inner.len -= n;
self.inner.head = self.inner.to_physical_idx(n);
ptr::drop_in_place(a);
ptr::drop_in_place(b);
```
That seems to be what you're suggesting (except that to work properly it would also need a drop guard), but it's also exactly the same as what `Drain::drop` does for the range `..n`. In this case I opted for the simpler implementation, though I also considered adding a `truncate_front` method which would do the special case of `self.drain(..n)` as efficiently as possible.